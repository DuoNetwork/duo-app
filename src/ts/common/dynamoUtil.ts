import { DynamoUtil } from '@finbook/duo-market-data';

let config = {};
try {
	config = require(`../keys/aws.ui.${__KOVAN__ ? 'dev' : 'live'}.json`);
} catch (error) {
	console.log(error);
}

const dynamoUtil = new DynamoUtil(config, !__KOVAN__, '');

export default dynamoUtil;
