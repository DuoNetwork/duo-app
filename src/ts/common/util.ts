import { IContractPrice } from '@finbook/duo-contract-wrapper';
import { IPriceStatus, IStatus } from '@finbook/duo-market-data';
import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import { ISourceData } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public formatTime(e: number) {
		return moment(e).format('YYYY-MM-DD HH:mm:ss');
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<IContractPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gemini: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gdax: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
