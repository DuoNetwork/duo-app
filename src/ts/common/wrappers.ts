import {
	Constants as WrapperConstants,
	DualClassWrapper,
	EsplanadeWrapper,
	ICustodianWrappers,
	MagiWrapper,
	StakeWrapper,
	Web3Wrapper
} from '@finbook/duo-contract-wrapper';

export const web3Wrapper = new Web3Wrapper(
	window,
	(__KOVAN__ ? WrapperConstants.PROVIDER_INFURA_KOVAN : WrapperConstants.PROVIDER_INFURA_MAIN).replace('/v3', ''),
	'',
	!__KOVAN__
);
export const dualClassWrappers: ICustodianWrappers = {};
for (const type in web3Wrapper.contractAddresses.Custodians) {
	dualClassWrappers[type] = {};
	for (const tenor in web3Wrapper.contractAddresses.Custodians[type])
		dualClassWrappers[type][tenor] = new DualClassWrapper(
			web3Wrapper,
			web3Wrapper.contractAddresses.Custodians[type][tenor].custodian.address
		);
}

export const getDualClassWrapperByTypeTenor = (type: string, tenor: string) => {
	if (dualClassWrappers[type] && dualClassWrappers[type][tenor])
		return dualClassWrappers[type][tenor];
	return dualClassWrappers[WrapperConstants.BEETHOVEN][WrapperConstants.TENOR_PPT];
};

export const getDualClassAddressByTypeTenor = (type: string, tenor: string) => {
	if (web3Wrapper.contractAddresses.Custodians[type][tenor])
		return web3Wrapper.contractAddresses.Custodians[type][tenor];
	return web3Wrapper.contractAddresses.Custodians[WrapperConstants.BEETHOVEN][
		WrapperConstants.TENOR_PPT
	];
};

export const magiWrapper = new MagiWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Oracles[0].address
);

export const esplanadeWrapper = new EsplanadeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.MultiSigManagers[0].address
);

export const stakeWrapper0 = new StakeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Stakes[0].address
);

export const stakeWrapper1 = new StakeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Stakes[1].address
);

export const stakeWrappers = [stakeWrapper0, stakeWrapper1];

export const calculateNav = DualClassWrapper.calculateNav;

export const getTokenInterestOrLeverage = DualClassWrapper.getTokenInterestOrLeverage;
