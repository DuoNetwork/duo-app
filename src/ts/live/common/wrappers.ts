import BeethovenWapper from '../../../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../../../duo-contract-wrapper/src/Web3Wrapper';
import infura from '../../keys/infura.json';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
export const web3Wrapper = new Web3Wrapper(window, '', provider, !__KOVAN__);
export const beethovenWapper = new BeethovenWapper(web3Wrapper);
