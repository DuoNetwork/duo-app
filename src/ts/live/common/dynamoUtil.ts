import AWS from 'aws-sdk';
import { ScanInput, ScanOutput } from 'aws-sdk/clients/dynamodb';
import devConfig from '../../keys/aws.ui.dev.json';
import liveConfig from '../../keys/aws.ui.live.json';
import { IPriceStatus, IStatus } from '../common/types';
import * as CST from './constants';

export class DynamoUtil {
	private ddb: AWS.DynamoDB;
	constructor() {
		// this.role = role;
		AWS.config.update(__DEV__ ? devConfig : liveConfig);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise((resolve, reject) => {
			this.ddb.scan(params, (err, data) => {
				if (err) reject(err);
				else resolve(data);
			});
		});
	}

	public async scanStatus() {
		return this.convertStatus(
			await this.scanData({
				TableName: __DEV__ ? CST.DB_AWS_STATUS_DEV : CST.DB_AWS_STATUS_LIVE
			})
		);
	}

	public convertStatus(status: ScanOutput): IStatus[] {
		if (!status.Items || !status.Items.length) return [];

		const output =  status.Items.map(d => {
			const process = d[CST.DB_ST_PROCESS].S || '';
			const timestamp = Number(d[CST.DB_ST_TS].N);
			if (process.startsWith('PRICE'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_TX_PRICE].N),
					volume: Number(d[CST.DB_TX_AMOUNT].N)
				} as IPriceStatus;
			else if (process.startsWith('FEED'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_HISTORY_PRICE].N),
					volume: Number(d[CST.DB_HISTORY_VOLUME].N)
				} as IPriceStatus;
			else
				return {
					process: process,
					timestamp: timestamp
				};
		});
		output.sort((a, b) => b.timestamp - a.timestamp);
		output.sort((a, b) => a.process.localeCompare(b.process));

		return output;
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
