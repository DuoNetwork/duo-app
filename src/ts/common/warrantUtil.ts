import { AttributeMap } from 'aws-sdk/clients/dynamodb';
import * as d3 from 'd3';
import moment from 'moment';
import * as CST from 'ts/common/constants';
import { web3Wrapper } from 'ts/common/wrappers';
import dynamoUtil from './dynamoUtil';
import * as StakingCST from './stakingCST';
//import { IAddressInfo, IStakingChild, IStakingNode } from './types';

class WarrantUtil {
	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	// public async insertReferralEntry(item: IReferral) {
	// 	const data: AttributeMap = {
	// 		address: { S: item.address.toLowerCase() },
	// 		referralCode: { S: item.referralCode.toLowerCase() },
	// 		signHash: { S: item.signHash },
	// 		updatedAt: { S: this.getUTCNowTimestamp() + '' }
	// 	};
	// 	const params = {
	// 		TableName: StakingCST.REFERRALTABLE,
	// 		Item: data
	// 	};

	// 	await dynamoUtil.insertData(params);
	// }

	public async getAddressInfo(account: string) {
		const params = {
			TableName: StakingCST.WARRENTTABLE,
			KeyConditionExpression: 'address = :address',
			ExpressionAttributeValues: {
				[':address']: { S: account }
			}
		};
		const data = await dynamoUtil.queryData(params);
		if (data.Count) {
			const roundStakingAmount = (data.Items as any)[0].roundStakingAmount.S.split(',').map(
				(d: any) => Number(d)
			);
			const roundReturn = (data.Items as any)[0].roundReturn.S.split(',').map((d: any) =>
				Number(d)
			);
			const boundETH = (data.Items as any)[0].boundETH.S.split(',');
			const rangeETH = [
				d3.format(',.2f')(Number(boundETH[0]) * (1 + Number(boundETH[1]))),
				d3.format(',.2f')(Number(boundETH[0]) * (1 - Number(boundETH[2])))
			];
			const addressInfo = {
				roundStakingAmount: roundStakingAmount,
				roundReturn: roundReturn,
				boundETH: rangeETH,
				settleETH: Number((data.Items as any)[0].settleETH.S),
				date: moment(Number((data.Items as any)[0].updatedAt.S)).format('YYYY-MM-DD')
			};
			return addressInfo;
		} else return {};
	}

	public async getCurrentRoundInfo(account: string) {
		const params = {
			TableName: StakingCST.UIEVENTTABLE,
			KeyConditionExpression: 'eventKey = :eventKey',
			ExpressionAttributeValues: {
				[':eventKey']: { S: account.toLowerCase() }
			}
		};
		const data = await dynamoUtil.queryData(params);
		if (data.Count) {
			const records = [];
			for (const item of data.Items as any) {
				const receipt = await web3Wrapper.getTransactionReceipt(item.transactionHash.S);
				const timeout =
					moment().valueOf() - Number(item.updateAt.S) > CST.PENDING_TX_TIMEOUT;
				const entry = {
					date: Number(item.updateAt.S),
					amount: Number(item.amount.S),
					txHash: item.transactionHash.S,
					status: receipt
						? receipt.status
							? 'mined'
							: 'error'
						: timeout
						? 'timeout'
						: 'pending'
				};
				records.push(entry);
			}
			records.sort((a, b) => a.date - b.date);
			return records;
		} else return [];
	}
	public async insetStakingEntry(item: any) {
		const data: AttributeMap = {
			eventKey: { S: item.address.toLowerCase() },
			amount: { S: item.amount },
			transactionHash: { S: item.txHash },
			updateAt: { S: this.getUTCNowTimestamp() + '' }
		};
		const params = {
			TableName: StakingCST.UIEVENTTABLE,
			Item: data
		};

		await dynamoUtil.insertData(params);
	}
	//public async getStakings(account: string) {}
}

const warrantUtil = new WarrantUtil();
export default warrantUtil;
