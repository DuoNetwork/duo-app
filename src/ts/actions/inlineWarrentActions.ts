import { IPrice } from '@finbook/duo-market-data';
//import moment from 'moment';
import * as CST from 'ts/common/constants';
import dynamoUtil from 'ts/common/dynamoUtil';
import { VoidThunkAction } from 'ts/common/types';
import util from 'ts/common/util';
import warrantUtil from 'ts/common/warrantUtil';
import {
	// 	stakeWrappers
	web3Wrapper
} from 'ts/common/wrappers';
export function balancesUpdate(duo: number) {
	return {
		type: CST.AC_STK_BALANCE,
		value: duo
	};
}

export function getBalances(): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		const duoBalance = await web3Wrapper.getErc20Balance(
			web3Wrapper.contractAddresses.DUO.address,
			account
		);
		dispatch(balancesUpdate(duoBalance));
	};
}

export function exchangePricesUpdate(prices: IPrice[]) {
	return {
		type: CST.AC_DCC_EX_PX,
		value: prices
	};
}

export function fetchExchangePrices(): VoidThunkAction {
	return async (dispatch) => {
		const source = 'kraken';
		const start = util.getUTCNowTimestamp() - 24 * 60 * 60000;
		dispatch(exchangePricesUpdate(await dynamoUtil.getPrices(source, 1, start, 0, 'ETH|USD')));
	};
}

export function boundariesUpdate(boundaries: number[]) {
	return {
		type: CST.AC_IW_BOUNDARIES,
		value: boundaries
	};
}

export function fetchBoundaries(): VoidThunkAction {
	return async (dispatch) => {
		const boundaries = await warrantUtil.getBoundaries();
		dispatch(boundariesUpdate(boundaries));
	};
}

export function currentRoundUpdate(currentRoundInfo: object[]) {
	return {
		type: CST.AC_IW_CURRENTROUND,
		value: currentRoundInfo
	};
}

export function fetchCurrentRoundInfo(): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		const currentRoundInfo = await warrantUtil.getCurrentRoundInfo(account);
		dispatch(currentRoundUpdate(currentRoundInfo));
	};
}

export function addressInfoUpdate(addressInfo: any) {
	return {
		type: CST.AC_IW_ADDRESSINFO,
		value: addressInfo
	};
}

export function fetchAddressInfo(): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		const addressInfo = await warrantUtil.getAddressInfo(account);
		dispatch(addressInfoUpdate(addressInfo));
	};
}
export function subscriptionUpdate(intervalId: number) {
	return {
		type: CST.AC_STK_SUB,
		id: intervalId
	};
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		dispatch(getBalances());
		dispatch(fetchCurrentRoundInfo());
		dispatch(fetchAddressInfo())
		dispatch(fetchBoundaries())
		dispatch(fetchExchangePrices());
	};
}

export function subscribe(): VoidThunkAction {
	return async dispatch => {
		dispatch(subscriptionUpdate(0));
		dispatch(refresh());
		dispatch(subscriptionUpdate(window.setInterval(() => dispatch(refresh()), 10000)));
	};
}

// export function refreshAdmin(): VoidThunkAction {
// 	return async dispatch => {
// 		await dispatch(getStates());
// 		dispatch(getAddresses());
// 	};
// }

// export function subscribeAdmin(): VoidThunkAction {
// 	return async dispatch => {
// 		dispatch(subscriptionUpdate());
// 		dispatch(refreshAdmin());
// 		dispatch(subscriptionUpdate(window.setInterval(() => dispatch(refreshAdmin()), 30000)));
// 	};
// }
