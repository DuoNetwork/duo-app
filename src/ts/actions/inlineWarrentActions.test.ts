// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import dynamoUtil from 'ts/common/dynamoUtil';
import util from 'ts/common/util';
import warrantUtil from 'ts/common/warrantUtil';
import { stakeV2Wrapper, web3Wrapper } from 'ts/common/wrappers';
import * as inlineWarrentActions from './inlineWarrentActions';

const stakeStore = configureMockStore([thunk]);

describe('actions', () => {
	stakeV2Wrapper.getStates = jest.fn(() =>
		Promise.resolve({
			canStake: false,
			canUnstake: false,
			lockMinTimeInSecond: 0,
			minStakeAmt: 0,
			maxStakePerOracle: 0,
			totalAwardsToDistribute: 0,
			stakingEnabled: false,
			totalRewardsToDistribute: 0
		})
	);
	stakeV2Wrapper.getAddresses = jest.fn(() =>
		Promise.resolve({
			operator: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0',
			priceFeedList: [
				'0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a',
				'0x002002812b42601Ae5026344F0395E68527bb0F8',
				'0x00476E55e02673B0E4D2B474071014D5a366Ed4E'
			],
			burnAddress: '0x61cA89CfC5E8099702e64e97D9b5FC457cf1d355',
			duoTokenAddress: '0x61cA89CfC5E8099702e64e97D9b5FC457cf1d355',
			uploader: '0x61cA89CfC5E8099702e64e97D9b5FC457cf1d355'
		})
	);
	stakeV2Wrapper.getOracleList = jest.fn(() =>
		Promise.resolve([
			'0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a',
			'0x002002812b42601Ae5026344F0395E68527bb0F8',
			'0x00476E55e02673B0E4D2B474071014D5a366Ed4E'
		])
	);
	stakeV2Wrapper.getUserStakes = jest.fn(() =>
		Promise.resolve({
			'0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a': [{ timestamp: 1234567890, amount: 123 }],
			'0x002002812b42601Ae5026344F0395E68527bb0F8': [{ timestamp: 1234567890, amount: 123 }]
		})
	);
	stakeV2Wrapper.getOracleStakes = jest.fn(() =>
		Promise.resolve({
			'0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a': 123,
			'0x002002812b42601Ae5026344F0395E68527bb0F8': 234
		})
	);
	stakeV2Wrapper.getUserReward = jest.fn(() => Promise.resolve(500));
	web3Wrapper.getErc20Balance = jest.fn(() => Promise.resolve(100));
	web3Wrapper.getErc20Allowance = jest.fn(() => Promise.resolve(100));

	test('balancesUpdate', () => {
		expect(inlineWarrentActions.balancesUpdate(123)).toMatchSnapshot();
	});

	test('getBalances', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(inlineWarrentActions.getBalances());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('allowanceUpdate', () => {
		expect(inlineWarrentActions.allowanceUpdate(123)).toMatchSnapshot();
	});

	test('getAllowance', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(inlineWarrentActions.getAllowance());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('userAwardUpdate', () => {
		expect(inlineWarrentActions.userAwardUpdate({} as any)).toMatchSnapshot();
	});

	test('getUserAward', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(inlineWarrentActions.getUserAward());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('exchangePricesUpdate', () => {
		expect(inlineWarrentActions.exchangePricesUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchExchangePrices', () => {
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = stakeStore({
			ui: {
				period: 5,
				source: 'test'
			}
		});
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair] as any)
		);
		store.dispatch(inlineWarrentActions.fetchExchangePrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('currentRoundUpdate', () => {
		expect(inlineWarrentActions.currentRoundUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchCurrentRoundInfo', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		warrantUtil.getCurrentRoundInfo = jest.fn((account: string) =>
			Promise.resolve([{ date: 1234567890, amount: 0, txHash: account, status: 'mined' }])
		);
		store.dispatch(inlineWarrentActions.fetchExchangePrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('addressInfoUpdate', () => {
		expect(inlineWarrentActions.addressInfoUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchAddressInfo', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		warrantUtil.getAddressInfo = jest.fn((account: string) =>
			Promise.resolve({
				roundStakingAmount: [100, 100],
				roundReturn: [100, 100],
				settleETH: 200.0,
				date: account,
				boundETH: [200, 200]
			})
		);
		store.dispatch(inlineWarrentActions.fetchExchangePrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('refresh', () => {
		util.getDates = jest.fn(() => ['1970-01-15']);
		const store: any = stakeStore({
			stake: {
				states: {},
				addresses: []
			},
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});

		store.dispatch(inlineWarrentActions.refresh());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('subscriptionUpdate', () => {
		expect(inlineWarrentActions.subscriptionUpdate(123)).toMatchSnapshot();
	});

	test('subscribe', () => {
		window.setInterval = jest.fn(() => 123);
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(inlineWarrentActions.subscribe());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect((window.setInterval as jest.Mock).mock.calls[0][1]).toMatchSnapshot();

				(window.setInterval as jest.Mock).mock.calls[0][0]();
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
