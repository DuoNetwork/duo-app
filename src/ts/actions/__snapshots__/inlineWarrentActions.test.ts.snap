// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions addressInfoUpdate 1`] = `
Object {
  "type": "iwaddressinfo",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions allowanceUpdate 1`] = `
Object {
  "type": "stakeAllowance",
  "value": 123,
}
`;

exports[`actions balancesUpdate 1`] = `
Object {
  "type": "stakeBalance",
  "value": 123,
}
`;

exports[`actions currentRoundUpdate 1`] = `
Object {
  "type": "iwcurrentround",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions exchangePricesUpdate 1`] = `
Object {
  "type": "dualClassExchangePrices",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions fetchAddressInfo 1`] = `
Array [
  Object {
    "type": "dualClassExchangePrices",
    "value": Array [
      "kraken",
      1,
      1148167890,
      0,
      "ETH|USD",
    ],
  },
]
`;

exports[`actions fetchCurrentRoundInfo 1`] = `
Array [
  Object {
    "type": "dualClassExchangePrices",
    "value": Array [
      "kraken",
      1,
      1148167890,
      0,
      "ETH|USD",
    ],
  },
]
`;

exports[`actions fetchExchangePrices 1`] = `
Array [
  Object {
    "type": "dualClassExchangePrices",
    "value": Array [
      "kraken",
      1,
      1148167890,
      0,
      "ETH|USD",
    ],
  },
]
`;

exports[`actions getAllowance 1`] = `
Array [
  Object {
    "type": "stakeAllowance",
    "value": 100,
  },
]
`;

exports[`actions getBalances 1`] = `
Array [
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
]
`;

exports[`actions getUserAward 1`] = `
Array [
  Object {
    "type": "userAward",
    "value": 500,
  },
]
`;

exports[`actions refresh 1`] = `
Array [
  Object {
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "stakingEnabled": false,
      "totalAwardsToDistribute": 0,
      "totalRewardsToDistribute": 0,
    },
  },
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
  Object {
    "type": "stakeAllowance",
    "value": 100,
  },
  Object {
    "type": "userAward",
    "value": 500,
  },
  Object {
    "type": "iwcurrentround",
    "value": Array [
      Object {
        "amount": 0,
        "date": 1234567890,
        "status": "mined",
        "txHash": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      },
    ],
  },
  Object {
    "type": "iwaddressinfo",
    "value": Object {
      "boundETH": Array [
        200,
        200,
      ],
      "date": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "roundReturn": Array [
        100,
        100,
      ],
      "roundStakingAmount": Array [
        100,
        100,
      ],
      "settleETH": 200,
    },
  },
  Object {
    "type": "dualClassExchangePrices",
    "value": Array [
      "kraken",
      1,
      1148167890,
      0,
      "ETH|USD",
    ],
  },
  Object {
    "type": "iwboundaries",
    "value": Array [
      0,
      0,
    ],
  },
]
`;

exports[`actions subscribe 1`] = `
Array [
  Object {
    "id": 0,
    "type": "stakeSubscription",
  },
  Object {
    "id": 123,
    "type": "stakeSubscription",
  },
  Object {
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "stakingEnabled": false,
      "totalAwardsToDistribute": 0,
      "totalRewardsToDistribute": 0,
    },
  },
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
  Object {
    "type": "stakeAllowance",
    "value": 100,
  },
  Object {
    "type": "userAward",
    "value": 500,
  },
  Object {
    "type": "iwcurrentround",
    "value": Array [
      Object {
        "amount": 0,
        "date": 1234567890,
        "status": "mined",
        "txHash": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      },
    ],
  },
  Object {
    "type": "iwaddressinfo",
    "value": Object {
      "boundETH": Array [
        200,
        200,
      ],
      "date": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "roundReturn": Array [
        100,
        100,
      ],
      "roundStakingAmount": Array [
        100,
        100,
      ],
      "settleETH": 200,
    },
  },
  Object {
    "type": "dualClassExchangePrices",
    "value": Array [
      "kraken",
      1,
      1148167890,
      0,
      "ETH|USD",
    ],
  },
  Object {
    "type": "iwboundaries",
    "value": Array [
      0,
      0,
    ],
  },
]
`;

exports[`actions subscribe 2`] = `10000`;

exports[`actions subscribe 3`] = `
Array [
  Object {
    "id": 0,
    "type": "stakeSubscription",
  },
  Object {
    "id": 123,
    "type": "stakeSubscription",
  },
  Object {
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "stakingEnabled": false,
      "totalAwardsToDistribute": 0,
      "totalRewardsToDistribute": 0,
    },
  },
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
  Object {
    "type": "stakeAllowance",
    "value": 100,
  },
  Object {
    "type": "userAward",
    "value": 500,
  },
  Object {
    "type": "iwcurrentround",
    "value": Array [
      Object {
        "amount": 0,
        "date": 1234567890,
        "status": "mined",
        "txHash": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      },
    ],
  },
  Object {
    "type": "iwaddressinfo",
    "value": Object {
      "boundETH": Array [
        200,
        200,
      ],
      "date": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "roundReturn": Array [
        100,
        100,
      ],
      "roundStakingAmount": Array [
        100,
        100,
      ],
      "settleETH": 200,
    },
  },
  Object {
    "type": "dualClassExchangePrices",
    "value": Array [
      "kraken",
      1,
      1148167890,
      0,
      "ETH|USD",
    ],
  },
  Object {
    "type": "iwboundaries",
    "value": Array [
      0,
      0,
    ],
  },
]
`;

exports[`actions subscriptionUpdate 1`] = `
Object {
  "id": 123,
  "type": "stakeSubscription",
}
`;

exports[`actions userAwardUpdate 1`] = `
Object {
  "type": "userAward",
  "value": Object {},
}
`;
